datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -258,6 +259,10 @@ impl CaseExpr {
     }
 }
 
+lazy_static! {
```

Comment:
We can't create temporary `Arc::new(NoOp::new())` objects in `children()`, but as `NoOp` is just a placeholder, that we need in `with_new_children()`, we can use 1 global instance. As `Arc`s can't be const, the simplest solution seemed to be using `lazy_static!`, but if someone has better idea please share.

Reply:
Can we skip pushing for the None cases? The user would know which indices refer to which expressions if needed. My concern is that this behavior could spread to expressions that have a flexible number of children.

Reply:
We need to restore the `CaseExpr` in `CaseExpr::with_new_children()` from the a `children: Vec<Arc<dyn PhysicalExpr>>`. If we don't push anyting for the 2 optional `CaseExpr.expr` and `CaseExpr.else_expr` fields, then I don't think there is a way to tell if the first element of `children` belongs to the optional `CaseExpr.expr` or to `CaseExpr.when_then_expr`.

Reply:
`with_new_children()` of `CaseExpr` says that: 
`// For physical CaseExpr, we do not allow modifying children size`.
That sentence tells me that we can also should not change the existence of expressions. Can we just insert the new children into the Some() ones in the element order?

Reply:
That's true, but please check how `NoOp`s are used there, if we don't have those `NoOp`s in `children` then how can we restore `expr`, `else_expr` and `when_then_expr`?

Reply:
In a world where `NoOp`'s are never used, I don't see a reason why this implementation wouldn't work.
```
        let expr = if self.expr().is_some() {
            Some(children[0].clone())
        } else {
            None
        };

        let else_expr = if self.expr().is_some() {
            Some(children[children.len() - 1].clone())
        } else {
            None
        };

        let branches = match (&expr, &else_expr) {
            (Some(_), Some(_)) => children[1..children.len() - 1].to_vec(),
            (Some(_), None) => children[1..children.len()].to_vec(),
            (None, Some(_)) => children[0..children.len() - 1].to_vec(),
            (None, None) => children[0..children.len()].to_vec(),
        };
        let mut when_then_expr: Vec<WhenThen> = vec![];
        for (prev, next) in branches.into_iter().tuples() {
            when_then_expr.push((prev, next));
        }
 ```
Of course, `NoOp`s are safer and don't require any assumptions, but I'm just curious about what I might be missing.

Reply:
Ah ok, I got it now.

Reply:
If we can assume that the number of children and their role can't change then we can do this: https://github.com/apache/datafusion/pull/10543/commits/b7eaa47c563f4e22f72dddae3d191fada74ffa92